lib/src/install.rs

```diff
@@ -1423,6 +1525,502 @@ impl BoundImages {
     }
 }
 
+pub(crate) fn open_composefs_repo(
+    rootfs_dir: &Dir,
+) -> Result<ComposefsRepository<Sha256HashValue>> {
+    ComposefsRepository::open_path(rootfs_dir, "composefs")
+        .context("Failed to open composefs repository")
+}
+
+async fn initialize_composefs_repository(
+    state: &State,
+    root_setup: &RootSetup,
+) -> Result<(Sha256Digest, impl FsVerityHashValue)> {
+    let rootfs_dir = &root_setup.physical_root;
+
+    rootfs_dir
+        .create_dir_all("composefs")
+        .context("Creating dir 'composefs'")?;
+
+    let repo = open_composefs_repo(rootfs_dir)?;
+
+    let OstreeExtImgRef {
+        name: image_name,
+        transport,
+    } = &state.source.imageref;
+
+    // transport's display is already of type "<transport_type>:"
+    composefs_oci_pull(&Arc::new(repo), &format!("{transport}{image_name}"), None).await
+}
+
+fn get_booted_bls() -> Result<BLSConfig> {
+    let cmdline = crate::kernel::parse_cmdline()?;
+    let booted = cmdline.iter().find_map(|x| x.strip_prefix("composefs="));
+
+    let Some(booted) = booted else {
+        anyhow::bail!("Failed to find composefs parameter in kernel cmdline");
+    };
+
+    for entry in std::fs::read_dir("/sysroot/boot/loader/entries")? {
+        let entry = entry?;
+
+        if !entry.file_name().as_str()?.ends_with(".conf") {
+            continue;
+        }
+
+        let bls = parse_bls_config(&std::fs::read_to_string(&entry.path())?)?;
+
+        if bls.options.contains(booted) {
+            return Ok(bls);
+        }
+    }
+
+    Err(anyhow::anyhow!("Booted BLS not found"))
+}
+
+pub(crate) enum BootSetupType<'a> {
+    /// For initial setup, i.e. install to-disk
+    Setup(&'a RootSetup),
+    /// For `bootc upgrade`
+    Upgrade,
+}
+
+#[context("Setting up BLS boot")]
+pub(crate) fn setup_composefs_bls_boot(
+    setup_type: BootSetupType,
+    // TODO: Make this generic
+    repo: ComposefsRepository<Sha256HashValue>,
+    id: &Sha256HashValue,
+    entry: ComposefsBootEntry<Sha256HashValue>,
+) -> Result<()> {
+    let id_hex = id.to_hex();
+
+    let (root_path, cmdline_refs) = match setup_type {
+        BootSetupType::Setup(root_setup) => {
+            // root_setup.kargs has [root=UUID=<UUID>, "rw"]
+            let mut cmdline_options = String::from(root_setup.kargs.join(" "));
+            cmdline_options.push_str(&format!(" composefs={id_hex}"));
```

Comment:
We should definitely be centralizing via e.g. `const` our references to this kernel commandline.